File: Go/백준/Gold/10713. 기차 여행/기차 여행.go
package main

import (
	"bufio"
	"os"
	"strconv"
)

type Cost struct {
	cash, card, IC int
}

func makePrefixSum(arr []int, N int) []int {
	prefixSum := make([]int, N+1)
	s := arr[0]
	for i := 1; i < len(arr); i++ {
		e := arr[i]
		if s > e {
			s, e = e, s
		}
		prefixSum[s]++
		prefixSum[e]--

		s = arr[i]
	}

	for i := 0; i < N; i++ {
		prefixSum[i+1] += prefixSum[i]
	}

	return prefixSum
}

func main() {
	bs.Split(bufio.ScanWords)
	defer bw.Flush()

	N, M := input(), input()
	arr := make([]int, M)
	for i := 0; i < M; i++ {
		arr[i] = input()
	}
	prefixSum := makePrefixSum(arr, N)

	costs := make([]Cost, N-1)
	for i := 0; i < N-1; i++ {
		cash, card, IC := input(), input(), input()
		costs[i] = Cost{cash, card, IC}
	}

	ans := 0
	for i := 1; i < N; i++ {
		if prefixSum[i] == 0 {
			continue
		}
		whenPayCash := prefixSum[i] * costs[i-1].cash
		whenPayCard := prefixSum[i]*costs[i-1].card + costs[i-1].IC

		ans += min(whenPayCash, whenPayCard)
	}
	bw.WriteString(strconv.Itoa(ans))
}

var (
	bs = bufio.NewScanner(os.Stdin)
	bw = bufio.NewWriter(os.Stdout)
)

func input() int {
	bs.Scan()
	num, _ := strconv.Atoi(bs.Text())
	return num
}

File: Python/백준/Gold/10713. 기차 여행/기차 여행.py
import sys; input = sys.stdin.readline


class Cost:
    def __init__(self, cash, card, IC):
        self.cash = cash
        self.card = card
        self.IC = IC


N,M = map(int,input().split())
arr = [*map(int,input().split())]
route = [sorted([arr[i], arr[i+1]]) for i in range(M-1)]

prefix_sum = [0] * (N+1)
for s,e in route:
    prefix_sum[s] += 1
    prefix_sum[e] -= 1

for i in range(N):
    prefix_sum[i+1] += prefix_sum[i]

cost = [Cost(*map(int,input().split())) for _ in range(N-1)]

ans = 0
for i in range(1,N):
    if not prefix_sum[i]: continue
    ans += min(prefix_sum[i] * cost[i-1].cash, prefix_sum[i] * cost[i-1].card + cost[i-1].IC)
print(ans)
